src/Orleans.Transactions/DistributedTM/TransactionAgent.cs

```diff
@@ -102,6 +105,9 @@ public Task<ITransactionInfo> StartTransaction(bool readOnly, TimeSpan timeout)
                 if (logger.IsEnabled(LogLevel.Debug))
                     logger.Debug($"{stopwatch.Elapsed.TotalMilliseconds:f2} timeout {transactionInfo.TransactionId} prepare responses");
 
+                foreach(var p in participants)
+                    p.Key.Abort(transactionInfo.TransactionId).Ignore();
```

Comment:
Can the agent make this call?  Isn't TM authoritative?  What if tm has committed transaction the, but this times out?

Reply:
This is the logic for read-only transaction. For those, there is no TM, instead the TA "commits" the transaction after it receives all the prepare responses. These responses are sent by the participants after they unlock, and after they persist the clock.